sqlitecluster/SQLiteNode.cpp

```diff
@@ -1665,51 +1716,57 @@ void SQLiteNode::_onMESSAGE(SQLitePeer* peer, const SData& message) {
             // APPROVE_TRANSACTION: Sent to the leader by a follower when it confirms it was able to begin a transaction and
             // is ready to commit. Note that this peer approves the transaction for use in the LEADING and STANDINGDOWN
             // update loop.
-            if (!message.isSet("ID")) {
-                STHROW("missing ID");
-            }
-            if (!message.isSet("NewCount")) {
-                STHROW("missing NewCount");
-            }
-            if (!message.isSet("NewHash")) {
-                STHROW("missing NewHash");
-            }
-            if (_state != SQLiteNodeState::LEADING && _state != SQLiteNodeState::STANDINGDOWN) {
-                STHROW("not leading");
-            }
-            SQLitePeer::Response response = SIEquals(message.methodLine, "APPROVE_TRANSACTION") ? SQLitePeer::Response::APPROVE : SQLitePeer::Response::DENY;
-            try {
-                // We ignore late approvals of commits that have already been finalized. They could have been committed
-                // already, in which case `_lastSentTransactionID` will have incremented, or they could have been rolled
-                // back due to a conflict, which would cuase them to have the wrong hash (the hash of the previous attempt
-                // at committing the transaction with this ID).
-                bool hashMatch = message["NewHash"] == _db.getUncommittedHash();
-                if (hashMatch && to_string(_lastSentTransactionID + 1) == message["ID"]) {
-                    if (message.calcU64("NewCount") != _db.getCommitCount() + 1) {
-                        STHROW("commit count mismatch. Expected: " + message["NewCount"] + ", but would actually be: "
-                              + to_string(_db.getCommitCount() + 1));
-                    }
-                    if (peer->permaFollower) {
-                        STHROW("permafollowers shouldn't approve/deny");
+
+            // If it's DENY or AsyncNotification is set means that it's not just a simple notification that the follower has some commit number.
+            // It's either a real DENY, or a real APPROVE of a quorum transaction.
+            if (SIEquals(message.methodLine, "DENY_TRANSACTION") || !message.isSet("AsyncNotification")) {
```

Comment:
this comment and code seem incongruent. The comment implies a DENY and AsyncNotification should be treated the same but the code is either DENY or NOT AsyncNotification.

Reply:
This is what we're aiming for:
```
// It's either a real DENY, or a real APPROVE of a quorum transaction.
```

This is the first condition:
```
if (SIEquals(message.methodLine, "DENY_TRANSACTION")
```

So that addresses `It's either a real DENY`

That leaves:
```
!message.isSet("AsyncNotification")
```

We've already established it's not a `DENY_TRANSACTION` or we wouldn't get to this evaluation, meaning that we're looking at an `APPROVE_TRANSACTION` message that is *not* an `AsyncNotification`, which should only be the case for QUORUM commands.

Reply:
should this comment then:

>If it's DENY or AsyncNotification is set means that it's not just a simple notification that the follower has some commit number.

be:

> If it's DENY or NOT AsyncNotification is set means that it's not just a simple notification that the follower has some commit number.

Reply:
Yes, you're right. The first line of the comment is wrong, the second line isn't.